src/util/Profiler.h

```diff
@@ -0,0 +1,112 @@
+#pragma once 
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "defines.h"
+
+#define ENABLE_PROFILER
+
+// Helper macros to generate unique* identifier
+// * Limitation: Only one PROFILER_X per line
+// TODO: Use __COUNTER__? It's not standard, but supported on most compilers
+#define _PROFILER_MERGE(A, B) A##B
+#define _PROFILER_LABEL(A) _PROFILER_MERGE(__profiler_, A)
+
+#ifdef ENABLE_PROFILER 
+	#define PROFILE_FUNC() auto _PROFILER_LABEL(__LINE__) = profiler->block(___func___);
+#else 
+	#define PROFILE_FUNC() block;
+#endif
+
+#ifdef ENABLE_PROFILER
+	#define PROFILE_BLOCK(name) auto _PROFILER_LABEL(__LINE__) = profiler->block(name)
+#else 
+	#define PROFILE_BLOCK(name) 
```

Comment:
See https://www.reddit.com/r/cpp/comments/g5f0ol/comment/fo42ngn?context=1 for example on how to get some magic function name detection to make this profiler more ergnomic for the reader.

Reply:
@jonesmz Feel free to correct/enhance it.

Reply:
It's going to be rather difficult for me to keep up with every commit or architectural change when they're merged in two days.

Most of my guidance for the project is in the project's discord, where I try to provide code examples, references to other projects that accomplish similar features, and similar.

I welcome you to have architectural and code discussions on discord https://discord.gg/XqTgAH

Reply:
I wont work on this profiler in near future as it is suitable for our needs now. Tatjam probably too. Your comment is longer than the actual code, please write it instead and you will free up Tatjam time so he can do other important things. Furthermore we can learn from your code how to optimize rest of it.

Reply:
I'm really confused what you are trying to communicate to me here.

That's not the role I have with this project. 

Either learn from the advice I provided, or don't, that's up to you.

Reply:
I'm new here. Don't know peoples roles in this project. Nice to meet you,